### PR TITLE
lazy load related / series / tone onward containers

### PR DESCRIPTION
**Pull request by mattosborn**
- added lazy load module to bind condition / load functions to window scroll
- lazy load related / series / tone onward containers if `lazyLoadOnwards` switch is on

// @gklopper

### CODE REVIEW

File: static/src/javascripts/projects/common/utils/lazy-loader.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    function addScrollingLazyLoader(el, distanceThreshold, loadFn) {
+        // calls `loadFn` when `scrollTop` is within `distanceThreshold` of `el`
+        var $el = bonzo(el),
+            conditionFn = function () {
+                var scrollTop = bonzo(document.body).scrollTop(),
+                    elOffset = $el.offset(),
+                    loadAfter = elOffset.top - distanceThreshold,
+                    loadBefore = elOffset.top + elOffset.height + distanceThreshold;
+                return scrollTop > loadAfter && scrollTop < loadBefore;
+            };
```
Comment (phamann) [inline static/src/javascripts/projects/common/utils/lazy-loader.js]: Let's discuss this function tomorrow before shipping. As I think we can make it slightly more performant. For instance you will be accessing `bonzo(document.body).scrollTop()` for every element in the array on each scroll event, which you should only need to read once.

Comment (phamann) [inline static/src/javascripts/projects/common/utils/lazy-loader.js]: Check out the Lonely Planets implementation of this here: https://github.com/lonelyplanet/rizzo/blob/master/app/assets/javascripts/lib/utils/proximity_loader.js **WARNING: CoffeeScript**

Comment (phamann) [inline static/src/javascripts/projects/common/utils/lazy-loader.js]: Also think @stephanfowler had some view on the best approach to proximity loading.

Comment (mattosborn) [inline static/src/javascripts/projects/common/utils/lazy-loader.js]: That Lonely Planet implementation has some limitations that wouldn't work for our use-case.

Reducing scrollTop calls was actually one of the reasons I abstracted like this, but hadn't finished it.

That said, we're only lazyloading 2 elements at most at the moment, and it is not likely to get beyond single figures. We're looking at a max of maybe 25-30 scrollTop calls every second. Considering jsPerf comes in at between 500k and 7m ops/sec for scrollTop I reckon the perf gain is negligible here.

Will have a look for a quick solution anyway.

File: static/src/javascripts/projects/common/utils/proximity-loader.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    doProximityLoadingThrottled = _.throttle(doProximityLoading, 200, {leading: false, trailing: true});
+    doProximityLoadingDebounced = _.debounce(doProximityLoading, 2000); // used on load for edge-case where user doesn't scroll
+
+    function addItem(conditionFn, loadFn) {
+        // calls `loadFn` when `conditionFn` is true
+        var item = {conditionFn: conditionFn, loadFn: loadFn};
+        items.push(item);
+        if (items.length === 1) {
+            mediator.on('window:scroll', doProximityLoadingThrottled);
```
Comment (stephanfowler) [inline static/src/javascripts/projects/common/utils/proximity-loader.js]: Sorry got to this too late. I've found using something `setInterval(doStuff, 500)` better than attaching to the scroll event. Doesn't need a debounce and can deal better with v fast scrolling.

**Pull request merged**